compiler/rustc_middle/src/ty/fast_reject.rs

```diff
@@ -42,7 +46,18 @@ where
     OpaqueSimplifiedType(D),
     FunctionSimplifiedType(usize),
     ParameterSimplifiedType,
-    ForeignSimplifiedType(DefId),
+}
+
+#[derive(PartialEq, Eq, Debug, Clone, Copy)]
+pub enum SimplifyParams {
+    Yes,
+    No,
+}
+
+#[derive(PartialEq, Eq, Debug, Clone, Copy)]
+pub enum StripReferences {
```

Comment:
I'm debating about these 2 boolean flags versus having "modes". It's not very obvious when one should use which value of these flags. I was thinking something like:

* Pattern -- treat the type as a pattern to be matched against (e.g. useful for the self type of an impl)
* TestTy -- treat the type as the type we are testing (e.g., the self type of the predicate to be proven)
* Diagnostic -- making diagnostic recommendations

but I'm not sure that's clearer! Maybe just a doc comment here explaining the role of each variable would be better.